peps/pep-0733.rst

```diff
@@ -0,0 +1,629 @@
+links to the individual issues. In addition, we included a section
+that aims to identify the different stakeholders of the C API,
+and the particular requirements that each of them has.
+
+
+C API Stakeholders
+==================
+
+As mentioned in the introduction, the C API was originally
+created as the internal interface between CPython's
+interpreter and the Python layer. It was later exposed as
+a way for third-party developers to extend and embed Python
+programs. Over the years, new types of stakeholders emerged,
+with different requirements and areas of focus. This section
+describes this complex state of affairs in terms of the
+actions that different stakeholders need to perform through
+the C API.
+
+Universal Actions for All Stakeholders
+--------------------------------------
+
+There are actions which are generic, and required by
+all types of API users:
+
+* Define functions and call them
+* Create classes and instantiate them
+* Create instances of builtin and user-defined types
+  and perform operations on them
+* Introspect objects, including types, instances, and functions
+* Raise and handle exceptions
+* Import modules
+* Manage threads
+* Manage sub-interpreters
+* Handle and send signals
+
+The following sections look at the unique requirements of various stakeholders.
+
+Extension Writers
+-----------------
+
+Extension writers are the traditional users of the C API. Their requirements
+are the universal actions listed above.
+
+Authors of Embedded Python Applications
+---------------------------------------
+
+Applications with an embedded Python interpreter. Examples are
+`Blender <https://docs.blender.org/api/current/info_overview.html>`__ and
+`OBS <https://obsproject.com/wiki/Getting-Started-With-OBS-Scripting>`__.
+
+They need to be able to:
+
+* Configure the interpreter (import paths, inittab, ``sys.argv``, etc.).
+* Interact with the execution model and program lifetime, including
+  clean interpreter shutdown and restart.
+* Represent complex data models in a way Python can use without
+  having to create deep copies.
+* Provide and import frozen modules.
+* Run multiple independent interpreters (in particular, when embedded
+  in a library that wants to avoid global effects).
+
+Alternative Python Implementations
+----------------------------------
+
+Alternative implementations of Python (such as
+`PyPy <https://www.pypy.org>`__,
+`GraalPy <https://www.graalvm.org/python/>`__,
+`IronPython <https://ironpython.net>`__,
+`RustPython <https://github.com/RustPython/RustPython>`__,
+`MicroPython <https://micropython.org>`__,
+and `Jython <https://www.jython.org>`__), may take
+very different approaches for the implementation of
+different subsystems. They need:
+
+* The API to be abstract and hide implementation details.
+* A specification of the API, ideally with a test suite
+  that ensures compatibility.
+* It would be nice to have an ABI that can be shared
+  across Python implementations.
```

Comment:
I'm not sure we are making this distinction between "Universal Actions" which must be compatible with CPython and "additional actions" that may not be.

Reply:
Let's not call them "Universal" then. (Or alternatively, let's officially decide that CPython is not the reference interpreter... but that's probably still a bit too controversial...)

We _ought_ to clearly distinguish between the intended goals ("universal Python API") and what we're evaluating here ("CPython's API", an imperfect superset of the universal Python API). There's just no way to define them as the same thing unless we immediately start ignoring that definition.

A really easy example is that we can't expect the initialization APIs to be identical between all implementations. So does that mean that initialization is not "universal" (my preference)? Or that we say initialization _is_ universal and everyone besides CPython is just incompatible with the universal Python API?

But my main point here is that every other group listed in this section has a list of jobs to be done, while this group has a list of aspects of the API that they hope to be true. It's inconsistent, and I think it would be better to list this group's jobs to be done (specifically, "implement the Universal Actions") before getting into their hopes and dreams for the ABI.

Reply:
"Universal" here was meant as "relevant to all stakeholders" (rather than a subset of the stakeholders). 

Regarding the requirements we list: they are not "what this stakeholder need to do" but rather "what this stakeholders needs from the C API". So "implement the Universal Actions" is not appropriate (or at least not specific enough). 

I see your point that the requirements of alternative implementations are of a different nature to those of people who use the API to implement something else. They are indeed a different type of stakeholder. An alternative implementation doesn't need to use the C API. It needs to know what it is so that it can implement it.  It doesn't need anything functional from the C API, it needs a good spec of what it does.

Reply:
> we can't expect the initialization APIs to be identical between all implementations

I assume you mean initialization of embedded (C)Python? I do not see any inherent reason why it cannot be universal. Java native interface (JNI) provides embedding API that is implemented by several different Java virtual machines.

(The initialization of extensions can be universal, which is demonstrated by HPy).

Reply:
Maybe universal was the wrong word. Would 'generic' or 'basic' be better?

Reply:
Maybe "Common" or "Fundamental" @iritkatriel. 'Generic' works too.

Reply:
"Common" is good. Thanks.